src/app/cli/src/commands/delete_command.rs

```diff
@@ -20,57 +21,133 @@ use crate::Interact;
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
+pub enum DeleteTarget {
+    Handles(Vec<DatasetHandle>),
+    Patterns(Vec<DatasetRefPattern>),
+}
+
 pub struct DeleteCommand {
-    interact: Arc<Interact>,
     dataset_repo: Arc<dyn DatasetRepository>,
     delete_dataset: Arc<dyn DeleteDatasetUseCase>,
-    dataset_ref_patterns: Vec<DatasetRefPattern>,
+    target: DeleteTarget,
     dependency_graph_service: Arc<dyn DependencyGraphService>,
+    push_service: Arc<dyn PushService>,
+    interact: Arc<Interact>,
     all: bool,
     recursive: bool,
 }
 
 impl DeleteCommand {
-    pub fn new<I>(
-        interact: Arc<Interact>,
+    pub fn new(
         dataset_repo: Arc<dyn DatasetRepository>,
         delete_dataset: Arc<dyn DeleteDatasetUseCase>,
-        dataset_ref_patterns: I,
+        target: DeleteTarget,
         dependency_graph_service: Arc<dyn DependencyGraphService>,
+        push_service: Arc<dyn PushService>,
+        interact: Arc<Interact>,
         all: bool,
         recursive: bool,
-    ) -> Self
-    where
-        I: IntoIterator<Item = DatasetRefPattern>,
-    {
+    ) -> Self {
         Self {
-            interact,
             dataset_repo,
             delete_dataset,
-            dataset_ref_patterns: dataset_ref_patterns.into_iter().collect(),
+            target,
             dependency_graph_service,
+            push_service,
+            interact,
             all,
             recursive,
         }
     }
+
+    async fn confirm_deletion(&self, dataset_handles: &[DatasetHandle]) -> Result<(), CLIError> {
+        for hdl in dataset_handles {
+            let statuses = self.push_service.status_multi(hdl).await?;
+
+            let mut out_of_sync = vec![];
+            let mut unknown = vec![];
+
+            for PushStatus {
+                remote,
+                check_result,
+            } in &statuses.statuses
+            {
+                match check_result {
+                    Ok(status) => match status {
+                        CompareChainsResult::LhsAhead { .. } => {
+                            out_of_sync.push((remote, "ahead of"));
+                        }
+                        CompareChainsResult::LhsBehind { .. } => {
+                            out_of_sync.push((remote, "behind"));
+                        }
+                        CompareChainsResult::Divergence { .. } => {
+                            out_of_sync.push((remote, "diverged from"));
+                        }
+                        _ => (),
+                    },
+                    Err(e) => unknown.push((remote, e)),
+                }
```

Comment:
The beauty of Rust allows customizing only what's different:

```
match check_result {
    Ok(status) => {
        let maybe_reason = match status {
            CompareChainsResult::LhsAhead => Some("ahead of"),
            ...
            _ => None
        };
        if let Some(reason) = maybe_reason {
            out_of_sync.push((remote, reason));
        }
    }
    ...
}
```

Reply:
done